Add fix to raise error when category value 'x' is not predefined but is assigned through df.loc[..]=x

chrispe opened the pull request:
- [X] closes #25383
- [X] tests passed, new test added
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

chrispe committed:
Add fix to raise error when category value is not predefined

chrispe92 commented:
To be honest I'm not quite happy with this solution. It looks a bit ugly to me. Do you think there's a nice way to include this fix specifically to the categorical module instead?

chrispe committed:
Fix linting

chrispe committed:
Added new test

simonjayhawkins commented:
> To be honest I'm not quite happy with my solution. Do you think there's a nicer way to include this fix specifically to the categorical module instead? It feels like that would require some major refactoring though. Any opinion?

see https://github.com/pandas-dev/pandas/issues/25383#issuecomment-465630713

from #25383, the expected behaviour is 

> Keep the categorical dtype if the added value is in the list of categories, throw an error/warning otherwise.

along with the test added here to confirm the an error raised, could also add test to confirm that unused categories are retained.

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Add test case for unused categories

pep8speaks commented:
Hello @chrispe92! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/arrays/categorical/test_indexing.py`](https://github.com/pandas-dev/pandas/blob/c627fa69ae9231f0355d3a40f31d993e6449bd1b/pandas/tests/arrays/categorical/test_indexing.py):

> [Line 251:61](https://github.com/pandas-dev/pandas/blob/c627fa69ae9231f0355d3a40f31d993e6449bd1b/pandas/tests/arrays/categorical/test_indexing.py#L251): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

chrispe committed:
Remove trailing whitespace

chrispe committed:
Fix linting

chrispe92 commented:
> > To be honest I'm not quite happy with my solution. Do you think there's a nicer way to include this fix specifically to the categorical module instead? It feels like that would require some major refactoring though. Any opinion?
> 
> see [#25383 (comment)](https://github.com/pandas-dev/pandas/issues/25383#issuecomment-465630713)
> 
> from #25383, the expected behaviour is
> 
> > Keep the categorical dtype if the added value is in the list of categories, throw an error/warning otherwise.
> 
> along with the test added here to confirm the an error raised, could also add test to confirm that unused categories are retained.

I've included a test for the use case you described. I managed to implement the assignment of invalid categorical value also with dataframe index expansion, but then of course it doesn't resolve the core issue. Should we still include it?

chrispe committed:
Fix linting

simonjayhawkins commented:
> I've included the test case you described. What are we looking to achieve with this PR at the end? My current changes resolve the issue of assigning a non-existing categorical value to a Categorical series, but then of course that doesn't resolve the core issue mentioned at #25383.

Thinking some more, I'm not sure if raising should be the expected behaviour. This happens at the array level, which is fine for the array as the dtype holds the categories and assigning values directly to the array that are not in the categories should raise.

however, for Series, if the dtype cannot hold an element a new array is created.

```
>>> ser = pd.Series([1, 2])
>>> ser
0    1
1    2
dtype: int64
>>>
>>> ser.values
array([1, 2], dtype=int64)
>>>
>>> id1 = id(ser.values)
>>> id1
2687613117264
>>>
>>> ser.loc[1] = 42
>>> ser
0     1
1    42
dtype: int64
>>>
>>> id(ser.values) == id1
True
>>>
>>> ser.loc[1] = "42"
>>> ser
0     1
1    42
dtype: object
>>>
>>> id(ser.values) == id1
False
>>>
```

For a performant operation, that does not create a new array, but updates the array in place and errors if the underlying array cannot hold the elements is Series.at. 

I would therefore expect Series.at to raise ValueError: Cannot setitem on a Categorical with a new...

but I would expect Series.loc to create a new array that can hold the new element with the correct categories. This issue needs further discussion IMO.

simonjayhawkins commented on pandas/core/generic.py:
```diff
@@ -5228,6 +5229,19 @@ def __setattr__(self, name: str, value) -> None:
         After regular attribute access, try setting the name
         This allows simpler access to columns for interactive use.
         """
+        # Fix for handling assignment to a Catagorical dtype
+        # with a category which is not predefined.
+        if isinstance(value, BlockManager) and is_categorical_dtype(self):
+            if len(value) > 0:
+                new_value = value.as_array()[-1]
+            else:
+                new_value = None
+            if not pd.isna(new_value) and new_value not in self.dtype.categories.values:
+                raise ValueError(
+                    "Cannot setitem on a Categorical with a new "
+                    "category, set the categories first"
+                )
+
```
not sure this is the correct location for the fix.

I suspect _setitem_with_indexer in pandas\core\indexing.py should be changed as this function is for setting values on a Series/DataFrame

not keen on seeing is_categorical_dtype in the code. would prefer using _can_hold_element but that maybe a more reaching change.

chrispe92 replied:
Thanks @simonjayhawkins! I'm not very familiar with the entire codebase, so your feedback has been very helpful to me. I'm currently having another look at it given your suggestions.

chrispe committed:
Remove temporary fix from generic.py

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe92 commented:
I'm a bit confused about the expected behaviour. In the current version this piece of code has the following result:
```python
>>> import pandas as pd
>>> from pandas import CategoricalDtype
>>> s_data = list("abcd")
>>> s_dtype = CategoricalDtype(list("abc"))
>>> s = pd.Series(s_data, dtype=s_dtype)
>>> s
0      a
1      b
2      c
3    NaN
dtype: category
Categories (3, object): [a, b, c]
`
Given that the dtype expects only values of (a,b,c), should we instead raise the same ValueError about non-invalid category value? If not, what's the logic behind the current behaviour?

chrispe committed:
First fix try through indexing.py

chrispe committed:
Fix lint

chrispe committed:
Fix import ordering

chrispe committed:
Fix Update

chrispe committed:
Fix lint

chrispe committed:
Include more related test cases

chrispe committed:
Fix linting

chrispe committed:
Update test_indexing.py

simonjayhawkins commented:
> Given that the `s` expects only values of `[a,b,c]` because of it's dtype, should we instead raise the same `ValueError` about non-invalid category value? If not, what's the logic behind the current behaviour? It's a bit contradictory to what we are trying to achieve with raising an exception (instead of replacing the new value with nan) when a non-existing category is inserted through index expansion.

so we have three possible behaviours?

1. raise if values not in categories
2. assign np.nan if values not in categories.
3. create a new array with new categories to hold new value.

and we have at least three operations where this applies

1. Series construction
2. assigning a new value to an existing Series element
3. Series expansion

The Series construction case is explicitly passing a dtype and so coercion of the values is expected. Assigning np.nan does seem reasonable but also raising is also a reasonable alternative. There are a couple of open issues regarding constructors, #24435 and #21311.

For this issue, we should maybe ignore the Series constructor for now, particularly since we are explicitly passing a dtype and therefore it is reasonable to allow a different behaviour and just focus on the two Series.\_\_setitem__ cases.

we need to step back slightly and consider a Series to to dictionary like, where the index is the keys, whereas the underlying CategorialArray is a sequence.

on master, the current behaviour is

setting an item for an existing key with a value not in the categories raises `ValueError: Cannot setitem on a Categorical with a new category, set the categories first`

as mentioned in https://github.com/pandas-dev/pandas/pull/34011#issuecomment-632150450, this behaviour for categorical is inconsistent with a numeric Series.

setting an item for an non-existing key with a value not in the categories, returns a Series with object dtype.

as mentioned in #25383, the current behaviour is unexpected.

issue #33952 was raised because of the inconsistency that currently exists, but in order to be consistent, we should also need to establish to correct behaviour.

> 1. raise if values not in categories

this is stated as the expected behaviour in both #33952 and #25383

> 2. assign np.nan if values not in categories.

This _may_ be OK in Series construction when a dtype is explicitly given, but for Series.\_\_setitem__, silently converting values to np.nan is probably not wanted.

> 3. create a new array with new categories to hold new value.

this would be consistent with numeric Series

jreback commented on pandas/core/indexing.py:
```diff
@@ -1838,7 +1855,16 @@ def _setitem_with_indexer_missing(self, indexer, value):
                         raise ValueError("cannot set a row with mismatched columns")
 
-                value = Series(value, index=self.obj.columns, name=indexer)
+                if len(set(self.obj.dtypes)) > 1:
```
this entire implementation should not be here, rather in Categorical indexing itself. Indexing is *already* very complicated and we are trying to remove things from core/indexing.py; this is dtype specific.

jreback replied:
pls address this comment

chrispe92 replied:
I removed the dtype specific code from `core/indexing.py`. However, some code changes still remain there, because I couldn't find an alternative way. The latest changes address the issue and so the `Categorical` dtype is now preserved:
```python
import pandas as pd

df = pd.DataFrame.from_dict({'reg': [0,1,2], 'cat':pd.Categorical(['a','b','b'], categories=['a','b','c','d'])})
print(df.dtypes)  # reg is int64, cat is categorical

df.loc[3] = (3, 'c')  # add row with categorical value that exist in categories
print(df.dtypes)  # reg is int64, cat is now still categorical (which is not the case in master)
```

Some additional points related to this PR changes:
- When an unknown category is appended through index expansion in a dataframe then that value is turned to nan.
```python
In [1]: import pandas as pd
   ...: from pandas import Categorical
   ...: df = pd.DataFrame({"int": [0, 1], "cat": pd.Categorical(["a", "b"], categories=["a", "b"])})
   ...: df.loc[3] = [3.3, "d"]
   ...: df
Out[1]:
   int  cat
0  0.0    a
1  1.0    b
3  3.3  NaN
```
- The same issue is not addressed for `pd.Series` in this PR. Which means that the dtype of the series changes to object when an unseen category is appended through index expansion.
```python
In [1]: import pandas as pd
   ...: from pandas import Categorical
   ...: ser = pd.Series(Categorical(["a", "b", "c"]))
   ...: ser.loc[3] = "d"
   ...: ser
Out[1]: 
0    a
1    b
2    c
3    d
dtype: object
```

chrispe92 commented:
> > Given that the `s` expects only values of `[a,b,c]` because of it's dtype, should we instead raise the same `ValueError` about non-invalid category value? If not, what's the logic behind the current behaviour? It's a bit contradictory to what we are trying to achieve with raising an exception (instead of replacing the new value with nan) when a non-existing category is inserted through index expansion.
> 
> so we have three possible behaviours?
> 
> 1. raise if values not in categories
> 2. assign np.nan if values not in categories.
> 3. create a new array with new categories to hold new value.
> 
> and we have at least three operations where this applies
> 
> 1. Series construction
> 2. assigning a new value to an existing Series element
> 3. Series expansion
> 
> The Series construction case is explicitly passing a dtype and so coercion of the values is expected. Assigning np.nan does seem reasonable but also raising is also a reasonable alternative. There are a couple of open issues regarding constructors, #24435 and #21311.
> 
> For this issue, we should maybe ignore the Series constructor for now, particularly since we are explicitly passing a dtype and therefore it is reasonable to allow a different behaviour and just focus on the two Series.__setitem__ cases.
> 
> we need to step back slightly and consider a Series to to dictionary like, where the index is the keys, whereas the underlying CategorialArray is a sequence.
> 
> on master, the current behaviour is
> 
> setting an item for an existing key with a value not in the categories raises `ValueError: Cannot setitem on a Categorical with a new category, set the categories first`
> 
> as mentioned in [#34011 (comment)](https://github.com/pandas-dev/pandas/pull/34011#issuecomment-632150450), this behaviour for categorical is inconsistent with a numeric Series.
> 
> setting an item for an non-existing key with a value not in the categories, returns a Series with object dtype.
> 
> as mentioned in #25383, the current behaviour is unexpected.
> 
> issue #33952 was raised because of the inconsistency that currently exists, but in order to be consistent, we should also need to establish to correct behaviour.
> 
> > 1. raise if values not in categories
> 
> this is stated as the expected behaviour in both #33952 and #25383
> 
> > 1. assign np.nan if values not in categories.
> 
> This _may_ be OK in Series construction when a dtype is explicitly given, but for Series.__setitem__, silently converting values to np.nan is probably not wanted.
> 
> > 1. create a new array with new categories to hold new value.
> 
> this would be consistent with numeric Series

When applying index expansion the `Categorical` constructor is called (through the `_setitem_with_indexer_missing`). As you mentioned, the constructor by default replaces the non-existing categories with `np.nan`. Therefore, it's not able to catch the case of including a non-existing categorical value through the index expansion, but to instead replace it with a missing value. And given that, I think it makes sense to stick to a single behaviour for all of the three cases you mentioned.

Index expansion makes it very difficult to identify for which of the solutions we should go for, since at any of the operations (series construction, expansion etc.) the same constructor function is called and in that phase you already luck the information of if that new object is used to construct a new Category series, or to expand an existing one.

dsaxton commented:
@chrispe92 Can you merge master to fix the conflict?

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
import missing dtypes function

chrispe committed:
Fix linting

chrispe92 commented:
Hi @dsaxton, it's done

jreback commented:
@chrispe92 also a number of comments.

chrispe committed:
Include requested changes

- Moved the tests for this PR to /tests/series/test_categorical.py
- Applied the code changes based on feedback

chrispe committed:
Fix import ordering/format

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Update test_categorical.py

chrispe committed:
Fix format

chrispe committed:
Remove commas

chrispe committed:
Update test_categorical.py

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Update solution

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Fix lint

chrispe committed:
Fix format issues

chrispe committed:
Update indexing.py

chrispe committed:
Update indexing.py

chrispe committed:
Update indexing.py

chrispe committed:
Update test_categorical.py

chrispe92 commented:
Made some new changes. Please check my last comment https://github.com/pandas-dev/pandas/pull/34011#discussion_r507150195 about them.

chrispe committed:
Update concat.py

chrispe committed:
Update cast.py

chrispe committed:
Update cast.py

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

jreback commented:
can  you merge master

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Update test_categorical.py

jreback commented:
this is *way* too complicated and a fragile code change. 

You might be able to not touch much of the indexing code and instead have concat correctly handling appending a value not in the category. These are de-facto the same case.

chrispe92 commented:
> this is _way_ too complicated and a fragile code change.
> 
> You might be able to not touch much of the indexing code and instead have concat correctly handling appending a value not in the category. These are de-facto the same case.

Ok, will try that instead and will keep you posted.

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Revert previous approach and include concat changes

chrispe committed:
Remove non-required convertion

chrispe committed:
Update concat.py

chrispe committed:
Update concat.py

chrispe92 commented:
@jreback : I have noticed a couple of tests in the master which contradict what we are looking to achieve through this MR.
Here is an example from `pandas.tests.reshape.concat.test_append_common`:

```python
# category + not-category => not-category
s1 = Series([1, 2, np.nan], dtype="category")
s2 = Series([2, 1, 2])

exp = Series([1, 2, np.nan, 2, 1, 2], dtype="object")
tm.assert_series_equal(pd.concat([s1, s2], ignore_index=True), exp)
```
The above test shows that preserving the categorical dtype is not expected when concatenating. 
So, why should this behaviour be different in the case of index expansion? 

Do we want to keep both behaviours on convert to one? Otherwise it would mean that arrays of a single element should be handled differently in the `concat_compat` function.

chrispe committed:
Update cast.py

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

jreback commented:
> @jreback : I have noticed a couple of tests in the master which partially contradict what we are looking to achieve with this MR (at least through the concat approach). Here is an example from `pandas.tests.reshape.concat.test_append_common`:
> 
> ```python
> # category + not-category => not-category
> s1 = Series([1, 2, np.nan], dtype="category")
> s2 = Series([2, 1, 2])
> 
> exp = Series([1, 2, np.nan, 2, 1, 2], dtype="object")
> tm.assert_series_equal(pd.concat([s1, s2], ignore_index=True), exp)
> ```
> 
> The above test shows that preserving the categorical dtype is not expected when concatenating.
> So, why should this behaviour be different in the case of index expansion (which at the end is also a concat
> with an array of a single element)?
> 
> Do we want to achieve both behaviours or convert only to one of them? If both, then it would mean that arrays of a single element should be handled differently in the `concat_compat` function (or revert back to the solution through indexing). Is that right, or am I missing something?

concatenation is different that indexing. I think raising on indexing is ok.

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

jreback commented:
@chrispe in order to merge this if you can respond to my comments

chrispe commented:
> @chrispe in order to merge this if you can respond to my comments

Hi @jreback, I'm not sure I completely understand your last comment about the raising. Can you elaborate a bit more?Thanks.

jreback commented:
if u r trying to add a non existent category (eg a value not in the categories) it should raise rather than coerce to object

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Add new version with raise

chrispe committed:
Add format fixes

chrispe committed:
Update test_categorical.py

chrispe commented on pandas/core/dtypes/concat.py:
```diff
@@ -108,7 +173,17 @@ def is_nonempty(x) -> bool:
+            # when an array of valid values is given (GH#25383)
+            if (
+                isinstance(to_concat[0], ExtensionArray)
+                and all(x.shape[0] == 1 for x in to_concat[1:])
```
I couldn't find a better way to detect when the `concat_compat` function is called through index expansion, so in cases like this:
```python
ser = pd.Series(Categorical(["a", "b", "c"]))
ser.loc[3] = "c"
```
With the latest commit we are raising a `ValueError` when an invalid value is added to the categorical through index expansion. it also enables the index expansion of a categorical of any `dtype`.

jreback commented on pandas/core/dtypes/concat.py:
```diff
@@ -108,7 +173,17 @@ def is_nonempty(x) -> bool:
+            if (
+                isinstance(to_concat[0], ExtensionArray)
+                and all(x.shape[0] == 1 for x in to_concat[1:])
+                and _can_cast_to_categorical(to_concat)
```
this a very complicated implementation. This should *all* be in `find_common_type`` , but should be *much* simpler that this. either the dtypes are the same or they are not. changing them is not in scope for this issue.

chrispe replied:
Ok, scope is more clear to me now. I will revert back to the previous approach and adapt it to raise on `dtype` mismatch.

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Update

chrispe committed:
Use prio_cat_dtype only for EAs

chrispe committed:
Revert usage of first_ea

chrispe committed:
Fix mypy errors

chrispe committed:
Use unique1d in _cast_to_common_type

chrispe committed:
Fix isort error

chrispe committed:
Renamed input variable for find_common_type

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Remove new argument in find_common_type

chrispe committed:
Add check to _get_common_dtype

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-category-index-issue-33952

chrispe committed:
Update dtypes.py

chrispe committed:
Update dtypes.py

chrispe committed:
Update dtypes.py

chrispe committed:
Update dtypes.py

chrispe committed:
Test

chrispe committed:
Add flag in get_common_type

chrispe committed:
Revert

chrispe committed:
Update dtypes.py

chrispe commented:
Hi @jreback. I did change my approach, but I does feel like I'm going in circles in terms of complexity, so I may need some further guidance. The thing is that the `find_common_type` does not take into account the values of the arrays, so we cannot conclude if we should raise an error given only their dtypes.

For example, this should raise an error:
```python
ser = pd.Series(Categorical(["a", "b", "c"]))
ser.loc[3] = "d"
# ValueError "Cannot setitem on a Categorical with a new category, set the categories first"
```
While this should be fine:
```python
ser = pd.Series(Categorical(["a", "b", "c"]))
ser.loc[3] = "a"
```

jreback commented:
closing as stale, if you want to continue working, please ping.
